arrayService.js

```diff
@@ -0,0 +1,173 @@
+    	}
+    	return arr.every(function(element){
+    		return element.isValid === true;
+    	});
+    }
+    
+    arrayService.isNodeValid = function(nodeId){
+        if (!arr[nodeId]){
+            throw new Error("value");
+        }
+        return arr[nodeId].isValid;
+    }
+    
+    arrayService.updateNode = function(node){
+        if (!node){
+            throw new Error("undefined node");
+        }
+        if (!node.id || !arr[+node.id.slice(4)]){
+            throw new Error("undefined node.id");
+        }
+        arr[+node.id.slice(4)] = {
+            isValid: isValidNumber(node.value),
+    		value: +node.value,
+    		node: node
+        }
+        return true;
+    }
+    
+    arrayService.swapNodes = function(firstNodeId, secondNodeId){
+        if (firstNodeId == undefined || secondNodeId == undefined){
+            throw new Error("undefined node.id");
+        }
+        if (!arr[firstNodeId] || ! arr[secondNodeId]){
+            throw new Error("undefined node.id");
+        }
+        
+        swapValues(firstNodeId, secondNodeId);
+        swapNodeValues(firstNodeId, secondNodeId);
+        return true;
+    }
+    
+    function swapValues(firstNodeId, secondNodeId){
+        var t = arr[firstNodeId].value;
+    	arr[firstNodeId].value = arr[secondNodeId].value;
+    	arr[secondNodeId].value = t;
+    }
+    
+    function swapNodeValues(firstNodeId, secondNodeId){
+        var t = arr[firstNodeId].node.value;
+    	arr[firstNodeId].node.value = arr[secondNodeId].node.value;
+    	arr[secondNodeId].node.value = t;
+    }
+    
+    arrayService.getLength = function(){
+        return arr.length;
+    }
+    
+    arrayService.sortStep = function(){
+        var result = {
+            activeNodeIds: [],
+            newStableNodeIds: [],
+            prevNodeIds: []
+        }
+        
+        if (currentArrPosition !== 0){
+            result.prevNodeIds.push(arr[currentArrPosition].node.id);
+            result.prevNodeIds.push(arr[currentArrPosition-1].node.id);
+        } else{
+            result.prevNodeIds.push(arr[currentStablePosition-1].node.id);
+        }
+        
+        if (currentArrPosition < currentStablePosition-1){
+            
+            result.activeNodeIds.push(arr[currentArrPosition].node.id);
+            result.activeNodeIds.push(arr[currentArrPosition+1].node.id);
+            
+    	    if (arr[currentArrPosition+1].value < arr[currentArrPosition].value) { 
+		        arrayService.swapNodes(currentArrPosition, currentArrPosition+1);
+	        }
+    	    currentArrPosition++;
+        } else{
+        	currentArrPosition = 0;
+    	    currentStablePosition--;
+    	    result.newStableNodeIds.push(arr[currentStablePosition].node.id);
+    	    if (currentStablePosition === 1){
+    		    currentStablePosition--;
+    	        result.newStableNodeIds.push(arr[currentStablePosition].node.id);
+    	   	}
+        }
+        return result;
+    }
+    
+    return arrayService;
+}
```

Comment:
Nice approach with arrayService. 
The code here is pretty easy readable and understandable.
Please pay more attention to indent. Some braces lost their indent. Especially at the end of file.